Fix assert and key error in --decisions flag

klausweinbauer opened the pull request:
This PR should fix issues with the --decisions flag generating ```assert count_false >= 0``` and ```KeyError: 0```. I encountered the same errors described in #637 and #640, which now works in my case. Let me know if there is still an issue.

Closes #637
Closes #640

latk commented on gcovr/decision_analysis.py:
```diff
@@ -156,11 +156,12 @@ def parse_one_line(self, lineno: int, code: str) -> None:
         if len(line_coverage.branches.items()) > 0:
             if _is_a_loop(code) or _is_a_oneline_branch(code):
                 if len(line_coverage.branches.items()) == 2:
+                    keys = list(line_coverage.branches.keys())
```
In practice this is correct, but using `sorted()` might make it clearer that we want the two smallest keys:

```suggestion
                    keys = sorted(line_coverage.branches)
```

klausweinbauer replied:
We actually want only two, not particularly the smallest, because the previous condition already guarantees only two branches. But if ```sorted()``` improves readability, I will change this.

Spacetown replied:
From Python 3.7 on the result is the same. In previous versions the inserted order isn't preserved and the branches can be in a different order. With `sorted` they are always ordered by the number as in the gcov output.

latk commented on gcovr/decision_analysis.py:
```diff
@@ -208,7 +209,7 @@ def continue_multiline_decision_analysis(self, lineno: int, code: str) -> None:
             last_decision_line_cov.decision = DecisionCoverageConditional(
                 exec_count,
-                last_decision_line_cov.count - exec_count,
+                max(last_decision_line_cov.count - exec_count, 0),
```
Yes this fixes the failing assertion, but if this count is negative that also means that our expectations about the control flow are wrong. In that case, do you think it's better to create a DecisionCoverageConditional(exec_count, 0) or should a DecisionCoverageUncheckable be generated?

klausweinbauer replied:
You are absolutely right. I thought about this but hadn't enough time to go into detail to find the cause for this misconception. It fixed the problem for now, and the generated decision results for my application seem reasonable. But I can go into more detail if you think this feature could still stand improvement.

Spacetown replied:
I prefer also a `DecisionCoverageUncheckable` instead of a count of 0.

latk reviewed (approved):
Thank you for working on this! As you've experienced, the decision coverage analysis is new and still has LOTS of bugs. Please continue to fix any issues you see. It's great that you identified these two details, though I have two comments.

Spacetown commented on gcovr/decision_analysis.py:
```diff
@@ -206,10 +207,17 @@ def continue_multiline_decision_analysis(self, lineno: int, code: str) -> None:
+                    delta_count,
+                )
+            else:
+                last_decision_line_cov.decision = DecisionCoverageUncheckable()
+                logger.debug(
```
Can you add simple test for this branch?

klausweinbauer commented:
It took me quite some time, but I managed to extract a reliably working test case from my project. I think it is as small as I can get it. It triggers the ```assert count_false >= 0``` error previous to this fix and now generates a ```DecisionCoverageUncheckable``` decision. This test could also help to find the original cause for this error. I think there is an issue with the execution count for certain lines; when calculated wrong, this could lead to negative values. 
I added a screenshot from the coverage of the test case. Line 7 is executed once according to Branch but counts two in Exec. Is this a bug, or did I miss something here?

I added a test reference output for gcc 9 generated by my native system, but unfortunately, it does not pass when run with docker. I am not quite sure what I did wrong. And is it possible to generate references for other versions with the provided docker
![fig](https://user-images.githubusercontent.com/51136324/181835495-2c7192c2-59d8-45b8-8ad2-5f9a8c83e171.png)
?

Spacetown commented:
> I added a test reference output for gcc 9 generated by my native system, but unfortunately, it does not pass when run with docker. I am not quite sure what I did wrong. And is it possible to generate references for other versions with the provided docker?

I've started the workflow.

How to update the reference is described in the contributors guide. You can run following command:
```bash
python3 -m nox --session 'docker_qa_compiler(all)' -- -k "decisions-neg-delta" --update_reference
```

Spacetown commented:
@klausweinbauer Can you add a commit to check if the pipeline is started?

klausweinbauer commented:
Added a test commit

Spacetown commented:
@klausweinbauer My fault, for new tests you should use `--generate_reference` instead of `--update_reference`.

Spacetown commented:
Data for clang Ang gcc-5 is missing.

klausweinbauer commented:
Okay, I will add this

Spacetown commented:
Thank you for working on this.
